modules/consensus/tendermint/verifier/src/verifier.rs

```diff
@@ -0,0 +1,321 @@
+				VerificationError::Invalid(
+					"Target header has no last_block_id (possibly genesis block)".to_string(),
+				)
+			})?
+			.hash;
+
+		if target_parent_hash.as_bytes() != trusted_state.finalized_header_hash {
+			return Err(VerificationError::Invalid(format!(
+				"Target header parent hash {:?} does not match trusted finalized hash {:?}",
+				target_parent_hash.as_bytes(),
+				trusted_state.finalized_header_hash
+			)));
+		}
+		return Ok(());
+	}
+
+	// Start from the target header and work backwards through ancestry
+	let mut expected_parent_hash = consensus_proof
+		.signed_header
+		.header
+		.last_block_id
+		.as_ref()
+		.ok_or_else(|| {
+			VerificationError::Invalid("Target header has no last_block_id".to_string())
+		})?
+		.hash;
+
+	for (i, header) in consensus_proof.ancestry.iter().enumerate().rev() {
+		let header_hash = header.header.hash();
+		if header_hash.as_bytes() != expected_parent_hash.as_bytes() {
+			return Err(VerificationError::Invalid(format!(
+				"Ancestry header {} hash mismatch: expected {:?}, got {:?}",
+				i,
+				expected_parent_hash.as_bytes(),
+				header_hash.as_bytes()
+			)));
+		}
+
+		expected_parent_hash = header
+			.header
+			.last_block_id
+			.as_ref()
+			.ok_or_else(|| {
+				VerificationError::Invalid(format!(
+					"Ancestry header {} has no last_block_id (possibly genesis block)",
+					i
+				))
+			})?
+			.hash;
+	}
+
+	if expected_parent_hash.as_bytes() != trusted_state.finalized_header_hash {
+		return Err(VerificationError::Invalid(format!(
+			"Ancestry chain does not connect to trusted finalized hash: expected {:?}, got {:?}",
+			trusted_state.finalized_header_hash,
+			expected_parent_hash.as_bytes()
+		)));
+	}
+
+	Ok(())
+}
+
+// Helper functions for type conversion
+
+fn convert_verification_options(
+	options: &VerificationOptions,
+	trusting_period: Duration,
+) -> Result<Options, VerificationError> {
+	let trust_threshold = TrustThresholdFraction::new(
+		options.trust_threshold_numerator,
+		options.trust_threshold_denominator,
+	)
+	.map_err(|e| VerificationError::ConversionError(format!("Invalid trust threshold: {}", e)))?;
+
+	Ok(Options { trust_threshold, trusting_period, clock_drift: options.clock_drift_duration() })
+}
+
+fn convert_timestamp(timestamp: u64) -> Result<Time, VerificationError> {
+	Time::from_unix_timestamp(timestamp as i64, 0)
+		.map_err(|e| VerificationError::Invalid(e.to_string()))
+}
+
+fn create_updated_trusted_state(
+	old_trusted_state: &TrustedState,
+	consensus_proof: &ConsensusProof,
+) -> Result<UpdatedTrustedState, VerificationError> {
+	// Check if validator set rotation is needed
+	let (validators, next_validators, new_next_validators_hash) = {
+		let signed_header_hash = &consensus_proof.signed_header.header.next_validators_hash;
+
+		// Only rotate if signed header has next_validators_hash and consensus proof has
+		// next_validators
+		if !signed_header_hash.is_empty() && consensus_proof.next_validators.is_some() {
```

Comment:
`signed_header_hash` is not empty and `consensus_proof.next_validators` is None, this function should return an error. We don't want the light client to be bricked.

Reply:
When `signed_header_hash` is empty, the `consensus_proof.next_validators` will be None. I had added this logic in the prover.

Reply:
You should remove this check here since it's already implemented in `extract_validators`, rotate the authority set if it's provided without any checks.